OCP-5999 rtph266depay

cfoch opened the pull request.

carlosfg-flu committed:
Implement rtph266pay

This is a first draft of a RTP payloder for H.266.

- [X]: parse NALUs (byte-stream)
- [X]: UP
- [X]: FU
- [X]: Mark
- [?]: Timestamps
- [X]: EOS/Drain/Flush
- [X]: _sink_event()
- [X]: PAUSED state
- [.]: XPS cache
- [ ]: Latency
- [ ]: Discont
- [?]: Delta

- [ ]: sprop
- [ ]: AP
- [ ]: DONL
- [ ]: stream-format={vvc1,???}
    - [ ]: codec_data
    - [ ]: parse NALUs (vvc1)
- [ ]: Trim trailing padding

NOTE: This commit is not intended to be merged upstream, but to be squashed
      first with future fixups.

carlosfg-flu committed:
rtph266pay: Add config-interval property

Keep a cache of all Parameter Sets received, then, send them when it's
appropriated according with the `config-interval` parameter.

Issue: OCP_6005

cfoch committed:
rtp: Add rtph266depay element

cfoch committed:
rtp: Add rtph266depay element

carlosfg-flu committed:
rtph266pay: Add support for Aggregation Packages

Add support for two types of aggregation:
* zero-latency: Only aggregate Parameter Set NALUs
* max: Try to aggregate entires NALUs

Issue: OCP_6005

carlosfg-flu committed:
rtph266pay: Amend errata

UNKOWN -> UNKNOWN

Issue: OCP_6005

carlosfg-flu commented:
Could you add the Jira ID to the PR title?

carlosfg-flu commented:
Still a draft?

carlosfg-flu commented:
Asan reports some leaks with this pipeline; but they can be in the decoder/payloader. I'm not sure.

```sh
GST_TRACERS='leaks(stack-traces-flags=full,filters=266,check-refs=true)' \
GST_DEBUG='GST_TRACER:7,rtph266*:5,flu*:5' \
gst-launch-1.0 \
	-v \
	videotestsrc \
	! fluh266enc \
		preset=faster \
	! rtph266pay \
	! rtph266depay \
	! fluh266dec \
	! videoconvert \
	! autovideosink \
	;
```

carlosfg-flu commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+
+  timestamp = GST_BUFFER_PTS (rtp->buffer);
+  marker = gst_rtp_buffer_get_marker (rtp);
+  GST_INFO_OBJECT (rtph266depay, "marker: %d", marker);
+
+  GST_INFO_OBJECT (rtph266depay,
+      "NAL header nal_unit_type %d, nuh_temporal_id_plus1 %d", nal_unit_type,
+      nuh_temporal_id_plus1);
```
`s/INFO/DEBUG/`

carlosfg-flu replied:
I really think these shouldn't be INFO messages but `GST_DEBUG_OBJECT()`. 

```
| 4 | INFO    | Logs all informational messages. These are typically used for  |
|   |         | events in the system that only happen once, or are important   |
|   |         | and rare enough to be logged at this level.
```

carlosfg-flu committed:
Merge pull request #37 from fluendo/rtph266pay

OCP-5966 rtph266pay

carlosfg-flu committed:
Merge pull request #37 from fluendo/rtph266pay

OCP-5966 rtph266pay

carlosfg-flu commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+    case GST_H266_NAL_AP:
+    {
+      // TODO: Payloader does not support Aggregation packets yet.
+      g_assert_not_reached ();
```
It's better to log an error and return `NULL` here. This path is correct; it's just not implemented. We must abort the pipeline in a controlled way.

Note that you should think that every `g_assert_* ()` can be disabled (deleted from the code) with `G_DISABLE_ASSERT`.

carlosfg-flu replied:
Also, I think this error path might deal with some resource liberation (like `rtp` or `payload`).

cfoch replied:
I added not_implemented goto statement.

carlosfg-flu commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+              "earlier packet. Dropping.");
+          gst_rtp_base_depayload_flush (depayload, FALSE);
+          gst_adapter_clear (rtph266depay->adapter);
+          return NULL;
```
Could `rtp` and/or `payload` have been leaked here?

cfoch replied:
- What do you mean by `payload`? Do you mean I miss to call gst_rtp_base_depayload_flush? 
- What do you mean by `rtp`? Do you mean freeing GstRTPBuffer?

carlosfg-flu replied:
> What do you mean by payload? Do you mean I miss to call gst_rtp_base_depayload_flush?

I meant the variable `payload`, but I've seen that you aren't the owner of the memory; it has the same lifetime as the `rtp` object.

> What do you mean by rtp? Do you mean freeing GstRTPBuffer?

Yes. But I've checked it, and I think it's not needed.

I've said nothing :slightly_smiling_face:

carlosfg-flu commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+              "stored.", rtph266depay->last_fu_seqnum, seqnum);
+          gst_rtp_base_depayload_flush (depayload, FALSE);
+          gst_adapter_clear (rtph266depay->adapter);
+          return NULL;
```
Are we liberating all resources here?

carlosfg-flu replied:
I was thinking about the `payload` and `rtp` variables, but as I mentioned in https://github.com/fluendo/gstreamer/pull/43#discussion_r1858812321 I no longer think there is a leak here.

cfoch commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+
+  g_assert (rtph266depay->alignment == GST_H266_ALIGNMENT_AU);
+
+  keyframe = NAL_TYPE_IS_KEY (nal_unit_type);
```
This is what rtph265depay does. I will check later.

cfoch commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+      break;
+    }
+    case GST_H266_NAL_FU:
+    {
```
I agree, but I prefer to leave it as that, so is closer similar to what GStreamer does, so easier to review, more chance to get approved... maybe...

cfoch commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+    default:
+      GST_DEBUG_OBJECT (rtph266depay, "Processing Single NAL Unit packet");
+      nalu_size = payload_len;
+      outsize = nalu_size + sizeof (sync_bytes);
+      outbuf = gst_buffer_new_and_alloc (outsize);
+
+      gst_buffer_map (outbuf, &map, GST_MAP_WRITE);
+
+      /* Assume byte-stream format. This is what template caps accepts right now. */
+      memcpy (map.data, sync_bytes, sizeof (sync_bytes));
+      memcpy (map.data + 4, payload, nalu_size);
+      gst_buffer_unmap (outbuf, &map);
```
There is not sync_bytes_buf or payload_buf. I would have to wrap them in a GstBuffer. Isn't it more problematic?

carlosfg-flu replied:
You only need 1 single `GstBuffer *sync_bytes_buf;`. You can allocate it in `gst_rtp_h266_depay_init()` and use it everywhere.

You can take a `GstBuffer *payload_buf` with `gst_rtp_buffer_get_payload_buffer()` (which will problaby avoid a memcpy)

cfoch commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+  GstStructure *structure = gst_caps_get_structure (caps, 0);
+  gint clock_rate;
+
+  if (!gst_structure_get_int (structure, "clock-rate", &clock_rate))
+    clock_rate = DEFAULT_CLOCK_RATE;
+  depayload->clock_rate = clock_rate;
```
I prefer to leave it as it is.

cfoch commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+        // rtph265depay->wait_start = FALSE;
+
+        /* reconstruct NAL header */
+        nal_header =
+            (((guint16) FUType) << 3) | (((guint16) nuh_layer_id) << 8) |
+            (guint16) nuh_temporal_id_plus1;
+        // GST_MEMDUMP_OBJECT (rtph266depay, "nal_header", (guint8 *) &nal_header,
+        //     sizeof (nal_header));
+
+        /* go back one byte so we can copy the payload + two bytes more in the front which
+         * will be overwritten by the nal_header
+         */
+        payload -= 1;
+        payload_len += 1;
+
+        nalu_size = payload_len;
+        outsize = nalu_size + sizeof (sync_bytes);
+        outbuf = gst_buffer_new_and_alloc (outsize);
+
+        gst_buffer_map (outbuf, &map, GST_MAP_WRITE);
+        memcpy (map.data, sync_bytes, sizeof (sync_bytes));
+        memcpy (map.data + sizeof (sync_bytes), payload, nalu_size);
+        map.data[sizeof (sync_bytes)] = nal_header >> 8;
+        map.data[sizeof (sync_bytes) + 1] = nal_header & 0xff;
+        gst_buffer_unmap (outbuf, &map);
```
What's the advantage with using _gst_buffer_append_? At the end all appended memories will be merged when calling gst_buffer_map. At the end, we are creating more GstBuffer structs (for the second argument of _gst_buffer_append_).

carlosfg-flu replied:
> What's the advantage of using gst_buffer_append?

Less code, easier to read.

> At the end all appended memories will be merged when calling gst_buffer_map. In the end, we are creating more GstBuffer structs (for the second argument of gst_buffer_append).

I'm not concerned about the performance here, but about the complexity. However, I'm not 100% sure `gst_buffer_map()` will be called. If I'm not wrong, `udpsink` doesn't use `gst_buffer_map()` (which indeed will imply a lot of memcpy to merge the memory). It's accessing each memory, and sending them one after another. This way, the memcpy for merging the memory is avoided.

cfoch commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+        outsize = nalu_size + sizeof (sync_bytes);
+        outbuf = gst_buffer_new_and_alloc (outsize);
+
+        gst_buffer_map (outbuf, &map, GST_MAP_WRITE);
```
?

carlosfg-flu replied:
Sorry. `gst_buffer_map()` can fail. Maybe you want to add an assert here.

cfoch commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+  /* add to adapter */
+  gst_buffer_unmap (nal, &map);
+  GST_DEBUG_OBJECT (depayload, "adding NAL to picture adapter");
+  gst_adapter_push (rtph266depay->picture_adapter, nal);
```
I think it makes sense, but, rtph264depay and rtph265depay is doing the same. In my opinion, it is better to fix this in bulk for all codecs, but that's another "issue".

cfoch commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+  GST_DEBUG_OBJECT (rtph266depay, "will allocate buffer of size %d", outsize);
+  outbuf = gst_rtp_h266_depay_allocate_output_buffer (rtph266depay, outsize);
+
+  if (outbuf == NULL)
+    return NULL;
```
It is possible if the allocation failed, no enough memory, for example. Unlikely, but possible. I will use G_UNLIKELY.

carlosfg-flu replied:
> It is possible if the allocation failed, no enough memory, for example.

You're right; `gst_buffer_new_allocate()` works differently than I thought. Glib always aborts when an allocation fails, AFAIK. I thought this allocation would follow the same rules, but it doesn't.

However, aborting is a good and usual way to handle when there is no more memory.

cfoch commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+        rtph266depay->current_fu_type = nal_unit_type;
+        rtph266depay->fu_timestamp = timestamp;
+        rtph266depay->last_fu_seqnum = seqnum;
+        // rtph265depay->wait_start = FALSE;
```
I will delete unused comment.

cfoch commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+  GST_DEBUG_OBJECT (depayload, "is discont %d",
+      GST_BUFFER_IS_DISCONT (rtp->buffer));
+
+  GST_FIXME_OBJECT (rtph266depay, "Assuming DONL field is not present");
```
I will leave as it is.

carlosfg-flu commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+        if (rtph266depay->current_fu_type == 0) {
+          /* previous FU packet missing start bit? */
+          g_critical ("missing FU start bit on an "
+              "earlier packet. Dropping.");
```
Pending

carlosfg-flu commented on subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c:
```diff
@@ -0,0 +1,674 @@
+          /* jump in sequence numbers within an FU is cause for discarding */
+          g_critical ("Jump in sequence numbers from "
+              "%u to %u within Fragmentation Unit. Data was lost, dropping "
+              "stored.", rtph266depay->last_fu_seqnum, seqnum);
```
Pending

carlosfg-flu reviewed (approved):
Approved. However, I strongly suggest changing the `INFO` messages to `DEBUG` and stopping using the `0b10101` compiler extension.

carlosfg-flu committed:
rtph266pay: XPS -> PS refactor

Remove X from XPS. Now it's only Package Set.

Issue: OCP_6005

carlosfg-flu committed:
rtph266pay: Add explanatory commentaries

Explain the more complex part of the algorithm, as it was proveen to be
hard to understand.

Issue: OCP_6005

carlosfg-flu committed:
rtph266pay: Update NALU_PTR_FORMAT and NALU_ARGS

Add missing Nalu fields to these parameters.

Issue: OCP_6005

carlosfg-flu committed:
rtph266pay: Add references to section numbers

Add commentaries with references to section numbers of the standards.

Issue: OCP_6005

carlosfg-flu committed:
Merge pull request #42 from fluendo/rtph266pay_xps

OCP-6005 rtph266pay: Add config-interval and AP
